### DISCUSSION TITLE
When the Consumer uses the asynchronous API, it leads to full CPU utilization.

### DISCUSSION THREAD

**Discussion by liweiyuan** [apache/pulsar#22105] [Q&A] [answered]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

pulsar-client: 3.x, 2.x

### Minimal reproduce step

1.use JAVA client , Consumer use async api
2. code：
`public class ConsumerTestCast {

    public static void main(String[] args) {

        try {
            PulsarClient client = PulsarClient.builder()
                    .serviceUrl("pulsar://192.168.2.213:6650")
                    .build();

            Consumer consumer = client.newConsumer()
                    .topic("Bosh-3")
                    .subscriptionName("TestGroup")
                    .ackTimeout(1, TimeUnit.SECONDS)
                    .subscribe();
            while(true){
                receiveAsync(consumer);
            }
        }catch (PulsarClientException e) {
            throw new RuntimeException(e);
        }
    }

    private static void receiveAsync(Consumer consumer) {
        CompletableFuture<Message> completableFuture = consumer.receiveAsync();
        completableFuture.whenComplete((msg,th) ->{
            if(msg != null){
                System.out.println("Message:" + new String(msg.getData()));
                try {
                    consumer.acknowledge(msg);
                } catch (PulsarClientException e) {
                    throw new RuntimeException(e);
                }
            }
            if(th != null){
                System.err.println("Err: "+ th.getMessage());
            }
        });
    }
}`

### What did you expect to see?

When using asynchronous APIs, there is no CPU consumption when there are no messages.

### What did you see instead?

Using the top command to check CPU usage, it was found that the CPU utilization is very high, maxing out the machine.

![2024-02-23 15-07-52屏幕截图](https://github.com/apache/pulsar/assets/26542900/f5654542-1d32-491e-8859-3b3bb576b76f)

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by lhotari** [comment, marked as answer]
@liweiyuan Your code example doesn't make sense for using the asynchronous API. This is essentially a tight loop:
```java
        while(true){
            receiveAsync(consumer);
        }
```

It seems that you should be using the listener feature instead.
Please check the docs about ["Create a consumer with a message listener"](https://pulsar.apache.org/docs/3.2.x/client-libraries-consumers/#create-a-consumer-with-a-message-listener).

If you are looking for a better asynchronous abstraction for the Pulsar Java client, I'd recommend using [pulsar-client-reactive](https://github.com/apache/pulsar-client-reactive) and/or [Spring Pulsar](https://spring.io/projects/spring-pulsar).

**Comment by liweiyuan** [comment]
ok
